akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
             closeThis(handler, elem)
             handOver(handler)
           } else {
-            // Drain into the void
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
+              if (eagerClose) completeStage()
```

Comment:
The problem is that there is no distinction between mid-stream cancelation and completion of a stream by fully consuming it, both are resulting as `onDownstreamFinish` callback (unless I'm missing something?).

This means that if we will complete stage on `onDownstreamFinish` of a substream when `eagerClose` flag is set, the stream will always terminate as soon as first substream completes (disregarding whether it was interrupted or fully consumed).

Here's an example:

```scala
Source(0 to 9000)
  .splitWhen(0 % 100 == 0, eagerClose = true)
  .concatSubstreams
```

Here each substream will be fully consumed, resulting in `onDownstreamFinish` after the last element of a substream is consumed. So if we complete the entire stage in `onDownstreamFinish` the stream will be interrupted after the first 100 elements.

The purpose of `eagerClose` is to avoid situations when substreams are cancelled without fully draining them, which currently leads to elements being drained automatically which may be unwelcome in case of big or even infinite streams or cases when items of a stream are tightly coupled with some resource consuming work (e.g. HTTP). However, if the substream has been fully consumed there is no point in tearing down the stage.